server/data_tiledb/tiledb_engine.py

```diff
@@ -0,0 +1,288 @@
+    def file_size(location):
+        return 0
+
+    @staticmethod
+    def open(location, args):
+        return TileDbEngine(location, args)
+
+    def get_location(self):
+        return self.url
+
+    def get_name(self):
+        return "cellxgene TileDb engine version"
+
+    def get_library_versions(self):
+        return dict(tiledb=tiledb.__version__)
+
+    def get_path(self, *urls):
+        return path_join(self.url, *urls)
+
+    def lsuri(self, uri):
+        """
+        given a URI, do a tiledb.ls but normalizing for all path weirdness:
+        * S3 URIs require trailing slash.  file: doesn't care.
+        * results on S3 *have* a trailing slash, Posix does not.
+
+        returns list of (absolute paths, type) *without* trailing slash
+        in the path.
+        """
+        def _cleanpath(p):
+            if p[-1] == '/':
+                return p[:-1]
+            else:
+                return p
+
+        if uri[-1] != '/':
+            uri += '/'
+
+        result = []
+        tiledb.ls(uri,
+                  lambda path, type: result.append((_cleanpath(path), type)),
+                  ctx=self.tiledb_ctx)
+        return result
+
+    @staticmethod
+    def isvalid(url):
+        """
+        Return True if this looks like a valid CXG, False if not.  Just a quick/cheap
+        test, not to be fully trusted.
+        """
+        if not tiledb.object_type(url) == "group":
+            return False
+        if not tiledb.object_type(path_join(url, "obs")) == "array":
+            return False
+        if not tiledb.object_type(path_join(url, "var")) == "array":
+            return False
+        if not tiledb.object_type(path_join(url, "X")) == "array":
+            return False
+        if not tiledb.object_type(path_join(url, "emb")) == "group":
+            return False
+        return True
+
+    def _validate_and_initialize(self):
+        if not self.isvalid(self.url):
+            raise RuntimeError(f"invalid tiledb dataset {self.url}")
+
+    def get_array(self, name, items=None):
+        p = self.get_path(name)
+        if items is None:
+            items = slice(None)
+        with tiledb.DenseArray(p, mode="r", ctx=self.tiledb_ctx) as A:
+            data = A[items]
+        return data
+
+    def get_embedding_array(self, ename, items=None):
+        return self.get_array(f"emb/{ename}", items)
+
+    def get_X_array(self, obs_items=None, var_items=None):
+        p = self.get_path("X")
+        if obs_items is None:
+            obs_items = slice(None)
+        if var_items is None:
+            var_items = slice(None)
+        # FIXME, the is a problem retrieving index sets from tiledb
+        with tiledb.DenseArray(p, mode="r", ctx=self.tiledb_ctx) as X:
+            tdata = X[:, :]
+            data = tdata[obs_items, var_items]
+        return data
+
+    def get_X_array_shape(self):
+        p = self.get_path("X")
+        with tiledb.DenseArray(p, mode="r", ctx=self.tiledb_ctx) as A:
```

Comment:
all of these repeated calls to tiledb.DenseArray will have a negative perf impact.  ideally you would do this once per HTTP request and pass the array 'A' around.   maybe the driver can obey the context manager protocol?

Reply:
alternatively, if the Driver instances are truly transient, we could simply create a cleanup protocol for them, ensuring that any cached resources have a chance to be cleaned up.

Reply:
I added a caching mechanism so that data adaptors can be reused between rest calls.
When a dataset is removed from the cache then a cleanup function is called to close all the arrays.